File: computing/interview/design-patterns/structural/bridge/java/BridgeDemo.java
package interview.designpatterns.structural.bridge;

interface Renderer {
    void renderCircle(double radius);
}

final class VectorRenderer implements Renderer {
    @Override
    public void renderCircle(double radius) {
        System.out.printf("Vector circle radius=%.2f%n", radius);
    }
}

final class RasterRenderer implements Renderer {
    @Override
    public void renderCircle(double radius) {
        System.out.printf("Raster circle radius=%.2f%n", radius);
    }
}

abstract class Shape {
    private final Renderer renderer;

    protected Shape(Renderer renderer) {
        this.renderer = renderer;
    }

    protected Renderer renderer() {
        return renderer;
    }

    abstract void draw();
}

final class Circle extends Shape {
    private final double radius;

    Circle(Renderer renderer, double radius) {
        super(renderer);
        this.radius = radius;
    }

    @Override
    void draw() {
        renderer().renderCircle(radius);
    }
}

public final class BridgeDemo {
    private BridgeDemo() {}

    public static void main(String[] args) {
        Shape vectorCircle = new Circle(new VectorRenderer(), 2.5);
        Shape rasterCircle = new Circle(new RasterRenderer(), 2.5);
        vectorCircle.draw();
        rasterCircle.draw();
    }
}

File: computing/interview/design-patterns/structural/bridge/csharp/BridgeDemo.cs
using System;

namespace Interview.DesignPatterns.Structural.Bridge
{
    public interface IRenderer
    {
        void RenderCircle(double radius);
    }

    public sealed class VectorRenderer : IRenderer
    {
        public void RenderCircle(double radius) => Console.WriteLine($"Vector circle radius={radius}");
    }

    public sealed class RasterRenderer : IRenderer
    {
        public void RenderCircle(double radius) => Console.WriteLine($"Raster circle radius={radius}");
    }

    public abstract class Shape
    {
        protected Shape(IRenderer renderer)
        {
            Renderer = renderer;
        }

        protected IRenderer Renderer { get; }
        public abstract void Draw();
    }

    public sealed class Circle : Shape
    {
        private readonly double _radius;

        public Circle(IRenderer renderer, double radius) : base(renderer)
        {
            _radius = radius;
        }

        public override void Draw()
        {
            Renderer.RenderCircle(_radius);
        }
    }

    public static class Demo
    {
        public static void Main()
        {
            var vectorRenderer = new VectorRenderer();
            var rasterRenderer = new RasterRenderer();

            Shape vectorCircle = new Circle(vectorRenderer, 2.5);
            Shape rasterCircle = new Circle(rasterRenderer, 2.5);

            vectorCircle.Draw();
            rasterCircle.Draw();
        }
    }
}
